pkg/driver/nydus/packer/packer.go

```diff
@@ -0,0 +1,304 @@
+		return "", nil, errors.Wrapf(err, "create work dir")
+	}
+
+	// Create a directory to store nydus blob file for every layer.
+	blobDirPath := path.Join(workDir, "blobs")
+	if err := os.MkdirAll(blobDirPath, 0755); err != nil {
+		return "", nil, errors.Wrapf(err, "create blob dir %s", blobDirPath)
+	}
+
+	// Create a directory to store nydus bootstrap file for every layer.
+	bootstrapDirPath := path.Join(workDir, "bootstraps")
+	if err := os.MkdirAll(bootstrapDirPath, 0755); err != nil {
+		return "", nil, errors.Wrapf(err, "create bootstrap dir %s", bootstrapDirPath)
+	}
+
+	cleanup := func() error {
+		return os.RemoveAll(workDir)
+	}
+
+	return workDir, cleanup, nil
+}
+
+func (p *Packer) diffBuild(ctx context.Context, workDir string, layers []*BuildLayer, diffSkip *int) (*builder.Output, error) {
+	diffPaths := []string{}
+	diffHintPaths := []string{}
+
+	bootstrapDir := path.Join(workDir, "bootstraps")
+	blobDir := path.Join(workDir, "blobs")
+	outputJSONPath := path.Join(workDir, "output.json")
+
+	for _, layer := range layers {
+		diffPaths = append(diffPaths, layer.diffPath)
+		diffHintPaths = append(diffHintPaths, layer.diffHintPath)
+	}
+
+	var parentBootstrapPath *string
+	if diffSkip != nil {
+		// Found nydus bootstrap cache, unpack targz and use it as parent bootstrap.
+		_parentBootstrapPath := path.Join(bootstrapDir, "parent-bootstrap")
+		parentBootstrapPath = &_parentBootstrapPath
+		layer := layers[*diffSkip]
+		cachedBootstrap, _ := layer.GetCache(ctx, CompressionTypeBootstrap)
+		if cachedBootstrap == nil {
+			return nil, fmt.Errorf("can't find bootstrap cache")
+		}
+		ra, err := layer.ContentStore(ctx).ReaderAt(ctx, ocispec.Descriptor{
+			Digest: cachedBootstrap.Digest,
+			Size:   cachedBootstrap.Size,
+		})
+		if err != nil {
+			return nil, errors.Wrap(err, "read bootstrap from content store")
+		}
+		defer ra.Close()
+
+		cr := content.NewReader(ra)
+		if err := utils.UnpackFile(cr, utils.BootstrapFileNameInLayer, _parentBootstrapPath); err != nil {
+			return nil, errors.Wrap(err, "unpack nydus bootstrap")
+		}
+	}
+
+	build := builder.New(p.builderPath)
+	output, err := build.Run(builder.Option{
+		BootstrapDirPath:    bootstrapDir,
+		BlobDirPath:         blobDir,
+		ParentBootstrapPath: parentBootstrapPath,
+
+		DiffLayerPaths:     diffPaths,
+		DiffHintLayerPaths: diffHintPaths,
+		DiffSkipLayer:      diffSkip,
+
+		OutputJSONPath: outputJSONPath,
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "build layers %v %v", diffPaths, diffHintPaths)
+	}
+
+	return output, nil
+}
+
+func (p *Packer) Build(ctx context.Context, layers []SourceLayer) ([]Descriptor, error) {
+	var diffSkip *int
+	var parent *BuildLayer
+
+	descs := make([]Descriptor, len(layers))
+
+	// Find cache first, to skip layers that have been built.
+	buildLayers := []*BuildLayer{}
+	for idx := range layers {
+		layer := layers[idx]
+
+		// Find the layer cache.
+		cachedBootstrapDesc, _ := layer.GetCache(ctx, CompressionTypeBootstrap)
```

Comment:
error shouldn't be ignored or should have a comment at least.

Reply:
Added a comment.